knt-backend/internal/kntdb/receipt.go

```diff
@@ -0,0 +1,208 @@
+func getProductValue(entry PurchaseEntry) (int, error) {
+	value, err := getFirstSingleValue[int]("select price from product where id = ? and visibility = 1", entry.ProductId)
+	return value * entry.Amount, err
+}
+
+// Generated the receipt and stores it in the database
+func generateReceipt(transaction *sql.Tx, entries []PurchaseEntry) (int64, error) {
+	dataString, err := json.Marshal(entries)
+	if err != nil {
+		return 0, err
+	}
+
+	return addReceipt(transaction, string(dataString))
+}
+
+func addReceipt(transaction *sql.Tx, dataString string) (int64, error) {
+	receiptId, err := addToTransaction(transaction, "INSERT INTO receipts (data, timestamp) VALUES (?, datetime())", dataString)
+	if err != nil {
+		return 0, err
+	}
+
+	return receiptId, nil
+}
+
+// Generates the transaction and stores it in the database
+func generateTransaction(transaction *sql.Tx, userId int, startingBal int, deltaBal int, finalBal int, receiptId int64, ref string) error {
+	_, err := addToTransaction(transaction,
+		"INSERT INTO transactions (user_id, starting_balance, delta_balance, final_balance, receipt_id, ref) VALUES (?, ?, ?, ?, ?, ?)",
+		userId, startingBal, deltaBal, finalBal, receiptId, ref)
+	return err
+}
+
+// Sets the users balance to a specified ammount
+func setBalance(transaction *sql.Tx, userId int, balance int) error {
+	_, err := addToTransaction(transaction, "UPDATE user SET balance = ? WHERE id = ?",
+		balance, userId)
+	return err
+}
+
+// Function designed to calculate total ammounts for tax reasons
+// The taxes are stored in their own tables in the form of tax{year}
+func addTaxTotals(entries []PurchaseEntry, transaction *sql.Tx, cost int) error {
+	year := time.Now().Year()
+
+	//Ensure yearly tables existance
+	err := ensureTaxTableExists(transaction, year)
+	if err != nil {
+		return err
+	}
+	//Go through all the entries and apply the operation on all of them
+	for _, entry := range entries {
+		//check if entry exists
+		tax, err := getFirstEntry[TaxEntry]("SELECT * FROM tax where year = ? and product_id = ?",
+			year, entry.ProductId)
+		if err != nil {
+			return err
+		}
+		if tax.Id == 0 {
+			//The row for this product doesnt already exist in the table so create a new row for it
+			_, err = addToTransaction(transaction, "INSERT INTO tax (product_id, amount, totalCost, year) VALUES (?, ?, ?, ?)",
+				entry.ProductId, entry.Amount, cost, year)
+			if err != nil {
+				return err
+			}
+			continue
+		}
+		//The product exists update it
+		_, err = addToTransaction(transaction, "UPDATE tax SET amount = ?, totalCost = ? WHERE year = ? and id = ?",
+			tax.Amount+entry.Amount, tax.TotalCost+cost, year, tax.Id)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func ensureTaxTableExists(transaction *sql.Tx, year int) error {
+	_, err := addToTransaction(transaction, "CREATE TABLE IF NOT EXISTS tax (id INTEGER PRIMARY KEY AUTOINCREMENT, product_id INT, amount INT, totalCost INT, year INT)")
+	return err
+}
```

Comment:
This is **_extremely_** cursed. I'm genuinely **not** willing to accept this.
You are **computing** the total taxes (good), but you are storing it back in the database, as if it were a fixed value, which it IS NOT.
On top of that, you are not even re-creating the table from scratch. So if you run this again, you end up with what you had before + the taxes again?! So like 2x taxes?!

This is cursed beyond belief. You should avoid duplicate data! Just use an SQL query to get the SUM (amount) of all the purchases of a certain item type and then multiply that by the tax percentage?
This is the job of a query, not of stored data.

Also, do you even commit the transaction? For some reasons I have doubts this even works.